pulp_puppet_plugins/pulp_puppet/plugins/migrations/0005_puppet_module_name_change.py

```diff
@@ -0,0 +1,59 @@
+from gettext import gettext as _
+import logging
+
+from mongoengine import NotUniqueError, Q
+from pulp.server.controllers import repository as repo_controller
+from pulp.server.db import model
+
+from pulp_puppet.plugins.db.models import Module
+
+_log = logging.getLogger('pulp')
+
+
+def migrate(*args, **kwargs):
+    """
+    For each puppet module check and if needed update module name format.
+
+    There was a discrepancy in the way puppet module's name was stored in pulp, depending
+    if it was synced from filesystem or uploaded. This migration finds puppet module units
+    that have wrong format name and replaces it with a correct format name.
+    """
+
+    modules = Module.objects(Q(name__contains='/') | Q(name__contains='-'))
+    repos_to_rebuild = set()
+    for puppet_unit in modules:
+        try:
+            author, name = puppet_unit['name'].split("-", 1)
+        except ValueError:
+            # This is the forge format, but Puppet still allows it
+            author, name = puppet_unit['name'].split("/", 1)
+        try:
+            puppet_unit.name = name
+            puppet_unit.save()
+        except NotUniqueError:
+            # find all repos that have this unit
+            repos_with_unit = model.RepositoryContentUnit.objects.filter(unit_id=puppet_unit.id)
+            repos_to_rebuild.update(repos_with_unit)
+            # find unit with correct name
+            correct_unit = Module.objects.filter(name=name).first()
+            for repo in repos_with_unit:
+                # unassociate wrong unit
+                repo_controller.disassociate_units(repo, [puppet_unit])
+                # associate correct unit to the list of the repos
+                repo_controller.associate_single_unit(repo, correct_unit)
+
+    repo_list = []
+    for repo in repos_to_rebuild:
+        repo_obj = model.Repository.objects.get_repo_or_missing_resource(repo.repo_id)
+        repo_controller.rebuild_content_unit_counts(repo_obj)
+        repo_list.append(repo.repo_id)
+
+    repos_to_republish = model.Distributor.objects.filter(repo_id__in=repo_list,
+                                                          last_publish__ne=None)
+    # redirect output to file
+    f = open('/tmp/repos_to_republish.txt', 'w')
```

Comment:
To be more cross platform this should be switched to using the `tempfile` module. At the very least the path could be generated with [`gettempprefix()`](https://docs.python.org/2/library/tempfile.html#tempfile.gettempprefix) but you could also use the functions in `tempfile` to replace the `open()` call altogether which would likely be the best.